pkg/server/server.go

```diff
@@ -0,0 +1,184 @@
+		logger: logger.WithName("azdo-server"),
+		port:   port,
+		proxy:  &httputil.ReverseProxy{Director: director},
+		authz:  authz,
+	}
+}
+
+// ListenAndServe starts the HTTP server on the specified port
+func (a *AzdoServer) ListenAndServe(stopCh <-chan struct{}) {
+	a.logger.Info("Starting git proxy", "port", a.port)
+	router := mux.NewRouter()
+	router.HandleFunc("/readyz", readinessHandler(a.logger)).Methods("GET")
+	router.HandleFunc("/healthz", livenessHandler(a.logger)).Methods("GET")
+	router.PathPrefix("/").HandlerFunc(proxyHandler(a.logger, a.proxy, a.authz))
+	srv := &http.Server{Addr: fmt.Sprintf(":%s", a.port), Handler: router}
+
+	go func() {
+		if err := srv.ListenAndServe(); err != http.ErrServerClosed {
+			a.logger.Error(err, "azdo-proxy server crashed")
+			os.Exit(1)
+		}
+	}()
+
+	<-stopCh
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	if err := srv.Shutdown(ctx); err != nil {
+		a.logger.Error(err, "azdo-proxy server graceful shutdown failed")
+	} else {
+		a.logger.Info("azdo-proxy server stopped")
+	}
+}
+
+func proxyHandler(logger logr.Logger, p *httputil.ReverseProxy, authz auth.Authorization) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		token, err := tokenFromRequest(r)
+		if err != nil {
+			w.Header().Set("WWW-Authenticate", "Basic realm=\"Restricted\"")
+			http.Error(w, "Missing basic authentication", http.StatusUnauthorized)
+			return
+		}
+
+		// Check basic auth with local auth configuration
+		err = authz.IsPermitted(r.URL.EscapedPath(), token)
+		if err != nil {
+			logger.Error(err, "Received unauthorized request")
+			http.Error(w, "User not permitted", http.StatusForbidden)
+			return
+		}
+		pat, err := authz.TargetForToken(token)
+
+		// Forward request to destination server
+		logger.Info("Authenticated request", "path", r.URL.Path)
+		r.Header.Del("Authorization")
+		patB64 := base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("pat:%s", pat)))
+		r.Header.Add("Authorization", "Basic "+patB64)
+		p.ServeHTTP(w, r)
+	}
+}
+
+func readinessHandler(log logr.Logger) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		w.Header().Set("Content-Type", "application/json")
+		if _, err := w.Write([]byte("{\"status\": \"ok\"}")); err != nil {
+			log.Error(err, "Could not write response data")
+		}
+	}
+}
+
+func livenessHandler(log logr.Logger) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		w.Header().Set("Content-Type", "application/json")
+		if _, err := w.Write([]byte("{\"status\": \"ok\"}")); err != nil {
+			log.Error(err, "Could not write response data")
+		}
+	}
+}
+
+func tokenFromRequest(r *http.Request) (string, error) {
+	username, password, ok := r.BasicAuth()
+	if !ok {
+		return "", fmt.Errorf("basic auth not present")
+	}
+
```

Comment:
I would like to use `found` instead of `ok`, since it could be interpreted that we are validating if the auth is ok or not - but we are checking for it's presence

Reply:
```go
// BasicAuth returns the username and password provided in the request's
// Authorization header, if the request uses HTTP Basic Authentication.
// See RFC 2617, Section 2.
func (r *Request) BasicAuth() (username, password string, ok bool) {
```

I do not want to go away from what is a go standard.